src/module.c

```diff
@@ -500,20 +500,18 @@ JL_DLLEXPORT jl_value_t *jl_get_global(jl_module_t *m, jl_sym_t *var)
 JL_DLLEXPORT void jl_set_global(jl_module_t *m JL_ROOTING_ARGUMENT, jl_sym_t *var, jl_value_t *val JL_ROOTED_ARGUMENT)
 {
     jl_binding_t *bp = jl_get_binding_wr(m, var, 1);
-    if (!bp->constp) {
-        bp->value = val;
-        jl_gc_wb(m, val);
-    }
+    assert(!bp->constp);
+    bp->value = val;
+    jl_gc_wb(m, val);
 }
 
 JL_DLLEXPORT void jl_set_const(jl_module_t *m JL_ROOTING_ARGUMENT, jl_sym_t *var, jl_value_t *val JL_ROOTED_ARGUMENT)
 {
     jl_binding_t *bp = jl_get_binding_wr(m, var, 1);
-    if (!bp->constp) {
-        bp->value = val;
-        bp->constp = 1;
-        jl_gc_wb(m, val);
-    }
+    assert(!bp->constp);
```

Comment:
Hm, by doing the assert on the condition we now do enable overwriting a const on non-debug build, whereas before they would get a crash later. Small nit-pick, but if we do it this way I think I would prefer an error to keep the semantics.

Reply:
If we throw an error that would be different semantics.

Reply:
Then let's move the assert into the else branch?